// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.grpc_service";
option java_outer_classname = "GRPCServiceProto";
option objc_class_prefix = "HLW";

package grpc_service;

// The DAMA service definition.
service DAMA {

  rpc bid_server (Bid) returns (BiddingResult) {}

  rpc get_server_price (Connection) returns (Price) {}

  rpc get_server_id (Connection) returns (ServerID) {}

  rpc set_layers_assigned (Assignment) returns (ServerResponse) {}

  rpc start_discounting (Connection) returns (Bid) {}

  rpc ack_layer (AddDropLayer) returns (ServerResponse) {}

  rpc return_layer (AddDropLayer) returns (ServerResponse) {}

  rpc nack_layer (AddDropLayer) returns (ServerResponse) {}

  rpc infer_layer (Features) returns (Preds) {}
}

message Bid {
  string device_id = 1;
  float bid_value = 2;
  string layer = 3;
  float benefit = 4;
  bool success = 5;
}

message BiddingResult{
  string server_id = 1;
  bool Ack = 2;
  float price_value = 3;
  string Nack_layer = 4;
}

message Price{
  float price_value = 1;
}

message ServerID{
  string server_id = 1;
}

message ServerResponse{
  bool success = 1;
}

message Assignment {
  bool layers_assigned = 1;
  string layer_profits = 2;
  string layer_benefits = 3;
  string device_id=4;
}

message Connection{
  string msg = 1;
}

message AddDropLayer {
  string layer = 1;
  float profit = 2;
  float benefit = 3;
}

message Features{
  string inputs = 1;
  string DAG = 2;
  bytes model_layer = 3;
}

message Preds{
  string output = 1;
}

// Compilation
// python -m grpc_tools.protoc -I protos --python_out=. --grpc_python_out=. protos/grpc_service.proto